crates/matrix-sdk-crypto/src/identities/manager.rs

```diff
@@ -644,44 +682,57 @@ impl IdentityManager {
         Ok((changes, changed_identity))
     }
 
-    /// Get a key query request if one is needed.
+    /// Get a list of key query requests needed.
     ///
-    /// Returns a key query request if the client should query E2E keys,
-    /// otherwise None.
+    /// # Returns
+    ///
+    /// A list of pairs of `(request_id, request)`
     ///
     /// The response of a successful key query requests needs to be passed to
     /// the [`OlmMachine`] with the [`receive_keys_query_response`].
     ///
     /// [`OlmMachine`]: struct.OlmMachine.html
     /// [`receive_keys_query_response`]: #method.receive_keys_query_response
-    pub async fn users_for_key_query(&self) -> StoreResult<Vec<KeysQueryRequest>> {
-        let users = self.store.users_for_key_query().await?;
+    pub async fn users_for_key_query(
+        &self,
+    ) -> StoreResult<Vec<(OwnedTransactionId, KeysQueryRequest)>> {
+        // forget about any previous key queries in flight
+        *(self.keys_query_request_details.lock().await) = KeysQueryRequestDetails::default();
```

Comment:
> Shouldn't we hold on to this lock since we're doing an assignment later bellow as well?

Possibly. It's not really obvious what behaviour is expected with concurrent calls to `users_for_key_query`/`receive_keys_query_response`, but I can pull out the lock as I think it's probably clearer.

> Why is it useful to reset first to a default state, it doesn't seem like we do anything with the reset thing before it gets reassigned to the new value?

There are a few places that `users_for_key_query` can bail out midway - notably if there are no queries to make, but also if there is an error. My thinking was that in this case we'd want to clear the request details rather than keep the old ones.

Reply:
> There are a few places that users_for_key_query can bail out midway - notably if there are no queries to make, but also if there is an error. My thinking was that in this case we'd want to clear the request details rather than keep the old ones.

Right, makes sense.